### PR TITLE
feat: db source

### PR DESCRIPTION
**Pull request by whynowy**
Signed-off-by: Derek Wang <whynowy@gmail.com>

TODO: add doc

mysql> select * from argo_dataflow_offsets;
+------------------------------------------------------------------+-------------------------------------------------------------+--------+
| uid                                                              | remark                                                      | offset |
+------------------------------------------------------------------+-------------------------------------------------------------+--------+
| 4e34996132d05b96ee2f0437889b7231a76cc905db82258e39c9362c739edfd8 | default.argo-dataflow-system.db-source.main.sources.default | 3      |
+------------------------------------------------------------------+-------------------------------------------------------------+--------+
1 row in set (0.00 sec)

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -22,6 +22,8 @@ require (
 	github.com/huandu/xstrings v1.3.2 // indirect
 	github.com/juju/fslock v0.0.0-20160525022230-4d5c94c67b4b
 	github.com/klauspost/compress v1.13.1 // indirect
+	github.com/lib/pq v1.10.2
```
Comment (alexec) [inline go.mod:25]: License?

File: runner/sidecar/source/db/db.go [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+	sharedutil "github.com/argoproj-labs/argo-dataflow/shared/util"
+	_ "github.com/go-sql-driver/mysql"
+	_ "github.com/lib/pq"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+var logger = sharedutil.NewLogger()
+
+var offsetTableSchema = `CREATE TABLE IF NOT EXISTS argo_dataflow_offsets (
```
Comment (alexec) [inline runner/sidecar/source/db/db.go:21]: Const?

File: runner/sidecar/source/db/db.go [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+	// update offset in db
+	go func() {
+		ticker := time.NewTicker(x.CommitInterval.Duration)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+				if offset != "" {
```
Comment (alexec) [inline runner/sidecar/source/db/db.go:104]: Can we only update once? Eg by setting offset to empty string.

File: runner/sidecar/source/db/db.go [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+	}, nil
+}
+
+func (d dbSource) Close() error {
+	return d.db.Close()
+}
+
+func getOffsetFromDB(ctx context.Context, db *sql.DB, tableName, consumerGroup string) (string, error) {
+	var offset string
+	if err := db.QueryRowContext(ctx,
```
Comment (alexec) [inline runner/sidecar/source/db/db.go:125]: Anyway to use a prepared query?

File: runner/sidecar/source/db/db.go [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+
+	columns, err := rows.Columns()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get table columns: %w", err)
+	}
+	count := len(columns)
+	result = make([]rowData, 0)
+	values := make([]interface{}, count)
+	valuePtrs := make([]interface{}, count)
+	for rows.Next() {
```
Comment (alexec) [inline runner/sidecar/source/db/db.go:180]: The result could be very large. Can we stream the data somehow?

**Review by alexec — COMMENTED** [review]
Some comments. I’d like to implement and architect the sidecar to have low memory requirements. So couple of changes:

* can you be more efficient with prepared statements?
* Can we avoid loading the whole record set into memory
* Should we limit the query to the commit size? The we can just commit inline once processed

**Comment by alexec** [comment]
I’m going to approve.I think change need marking, but they could be in a subsequent PR, or in this one.

File: go.mod (hunk 1 above)
Comment (whynowy) [reply go.mod]: removed postgresql support for now, figure it out later.

File: runner/sidecar/source/db/db.go (hunk 2 above)
Comment (whynowy) [reply runner/sidecar/source/db/db.go]: done.

File: runner/sidecar/source/db/db.go (hunk 3 above)
Comment (whynowy) [reply runner/sidecar/source/db/db.go:97]: it's only once.

File: runner/sidecar/source/db/db.go (hunk 4 above)
Comment (whynowy) [reply runner/sidecar/source/db/db.go]: done.

File: runner/sidecar/source/db/db.go (hunk 5 above)
Comment (whynowy) [reply runner/sidecar/source/db/db.go:181]: yes.

**Comment by whynowy** [comment]
> Some comments. I’d like to implement and architect the sidecar to have low memory requirements. So couple of changes:
> 
> * can you be more efficient with prepared statements?
> * Can we avoid loading the whole record set into memory
> * Should we limit the query to the commit size? The we can just commit inline once processed

Commit side/load size could be added later - considering different db has different sql for that.
Bad idea to commit inline once processed - if it fails at the middle due to a connection issue..
Kafka also commits every few seconds.

File: runner/sidecar/source/db/db.go [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+	sharedutil "github.com/argoproj-labs/argo-dataflow/shared/util"
+	_ "github.com/go-sql-driver/mysql"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+var logger = sharedutil.NewLogger()
+
+const offsetTableSchema = `CREATE TABLE IF NOT EXISTS argo_dataflow_offsets (
+	table_name VARCHAR(255) NOT NULL,
```
Comment (alexec) [inline runner/sidecar/source/db/db.go:21]: rather than table name, perhaps this could be uid, and uid is hash(cluster+namespace+pipeline+step+source)?

Comment (whynowy) [reply runner/sidecar/source/db/db.go:21]: That's the idea of another column `consumer_group`.

Comment (alexec) [reply runner/sidecar/source/db/db.go:21]: I think we should allow people to specify queries that select from many tables, this is flexible, and will be better performing and much simpler that having to side-load tables. Hence my thinking that we should not have table name here.

File: api/v1alpha1/db_source.go [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+
+type DBSource struct {
+	Database     `json:",inline" protobuf:"bytes,1,opt,name=database"`
+	Table        string `json:"table,omitempty" protobuf:"bytes,2,opt,name=table"`
+	OffsetColumn string `json:"offsetColumn,omitempty" protobuf:"bytes,3,opt,name=offsetColumn"`
+	// +kubebuilder:default="1s"
+	PollInterval metav1.Duration `json:"pollInterval,omitempty" protobuf:"bytes,4,opt,name=pollInterval"`
+	// +kubebuilder:default="5s"
+	CommitInterval metav1.Duration `json:"commitInterval,omitempty" protobuf:"bytes,5,opt,name=commitInterval"`
+	InitSchema     bool            `json:"initSchema,omitempty" protobuf:"bytes,6,opt,name=initSchema"`
```
Comment (alexec) [inline api/v1alpha1/db_source.go:15]: what is this for? maybe we just init always on leader start up?

File: runner/sidecar/source/db/db.go [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+	PRIMARY KEY(table_name, consumer_group)
+)`
+
+type rowData = map[string]interface{}
+
+type dbSource struct {
+	db *sql.DB
+}
+
+func New(ctx context.Context, secretInterface corev1.SecretInterface, clusterName, namespace, pipelineName, stepName string, replica int, sourceName string, x dfv1.DBSource, f source.Func) (source.Interface, error) {
```
Comment (alexec) [inline runner/sidecar/source/db/db.go:33]: In the future - we need a way, like S3, to load balance across the replicas. Probably only the lead replica (replica=0) should query and make HTTP callback to service with data. This allows us to, in future, process in parallel. 

I think we need shared code for this. I need to think more on the topic.

File: runner/sidecar/source/db/db.go (hunk 3 above)
Comment (alexec) [reply runner/sidecar/source/db/db.go:97]: what if we have less than one item every 5s? how many update do we do?

File: runner/sidecar/source/db/db.go [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+	defer func() { _ = stmt.Close() }()
+	var offset string
+	if err = stmt.QueryRowContext(ctx, tableName, consumerGroup).Scan(&offset); err != nil {
+		return "", err
+	}
+	return offset, nil
+}
+
+func updateOffset(ctx context.Context, db *sql.DB, tableName, consumerGroup, offset string) (int64, error) {
+	stmt, err := db.Prepare("update argo_dataflow_offsets set offset=? where table_name=? and consumer_group=?")
```
Comment (alexec) [inline runner/sidecar/source/db/db.go:130]: could have `and offset <> ?`

File: api/v1alpha1/db_source.go (hunk 7 above)
Comment (whynowy) [reply api/v1alpha1/db_source.go:15]: think about the case you are now allowed to have the permission to create tables and your admin helped you set up.

File: runner/sidecar/source/db/db.go (hunk 9 above)
Comment (whynowy) [reply runner/sidecar/source/db/db.go]: added.

File: runner/sidecar/source/db/db.go (hunk 6 above)
Comment (whynowy) [reply runner/sidecar/source/db/db.go]: done.

**Review by alexec — APPROVED** [review]

**Pull request merged**